### ISSUE TITLE
Медленная работа агрегации

### ISSUE THREAD

**Issue by rkomissarenko** [intaro/pinboard#126]
Заметил, что запускаемый каждые 15 минут скрипт агрегации данных (/pinboard/console aggregate) выполняется по времени как раз примерно эти же 15 минут.
Начал разбираться. Дошел до файла /root/pinboard/src/Pinboard/Command/AggregateCommand.php.
Увидел запрос на вставку данных в ipm_report_2_by_hostname_and_server. Он склеивается в один для всех серверов и выполняется. У меня он выполняется 11 минут.
Бер

**Comment by muxx** [comment]
А какие объемы данных?

**Comment by muxx** [comment]
Может быть, еще требуется посмотреть в сторону тюнинга innodb.

**Comment by rkomissarenko** [comment]
Обновил стартовый топик, ответив на вопрос.

**Comment by rkomissarenko** [comment]
> Может быть, еще требуется посмотреть в сторону тюнинга innodb.

В данном узком месте работа идёт с таблицей типа PINBA. Влияют ли настройки innodb на неё? MySQL оттюнингован. Понятно, что до совершенства далеко, но есть сомнения, что тюнинг решит данную ситуацию.

Первое, что приходит в голову - запускать агрегацию почаще. И поменьше держать данных в request. Есть ли какие-то другие способы?

Что-то крутить и править в структуре БД (менять типы таблиц, значения по умолчанию и т. д.) я не решаюсь, т. к. не знаю логику работы с ней.

**Comment by muxx** [comment]
Да, innodb – это только по части таблиц Pinboard. Таблицы pinba лежат в памяти, для них не предусмотрено тюннинга.. Аггрегацию лично я делаю раз в 10 минут, но у меня на порядок меньше данных. А не понятно, что тормозит, выборка или вставка?

**Comment by rkomissarenko** [comment]
> А не понятно, что тормозит, выборка или вставка?

Тормозит выборка из-за отсутствия индексов в таблице request. Практически уверен, были бы индексы, проблемы бы не было.

Я указал в первом сообщении фрагмент скрипта с подзапросом (квадратичная сложность). Он рассчитывается за 3-6 секунд. В одном запросе на вставку - 12 подзапросов. А всего около 90 insert-ов (на моей базе).

Решение - это не допускать разрастания таблицы request больше... сотни тысяч строк. Либо делать несколько pinba-серверов. Но мне кажется можно выжать гораздо большую производительность для большого объёма данных.

**Comment by muxx** [comment]
Таблица request в памяти и обновляется в реалтайме. Индексы будут дольше перестраиваться.

**Comment by muxx** [comment]
Я только вижу перенести это на сторону pinba. У нее есть возможность построения отчетов по персентилям, т.е. она в реалтайме будет и аггрегированные данные расчитывать, а pinboard будет только выбирать их и складывать в свои таблицы. На момент разработки Pinboard их еще не было, но надо смотреть, всего ли там хватает, чтобы мигрировать.

**Comment by rkomissarenko** [comment]
Согласен с вами. Будем ожидать новых версий pinboard :)

**Comment by rkomissarenko** [comment]
Поковырялся. Увидел ещё 2 варианта решения проблемы.
1. Дампить необходимые данные из таблицы request в свою таблицу, где уже выставлены все индексы. Запросы агрегации делать уже в новой таблице. После всех действий делаем TRUNCATE.
2. Новую таблицу не заводить, перенести логику из SQL на PHP (сортировки, группировки).
